### PR TITLE
Emit in_spec and out_spec, relax some asserts, intermediate step towards non-modified callable

### PR DESCRIPTION
**Pull request by voznesenskym**

### CODE REVIEW

File: torchdynamo/eval_frame.py [hunk 1]
```diff
@@ -279,16 +279,28 @@ def dynamo_normalization_capturing_compiler(
...
+            sig_type = signature_types[idx]
+            in_type = in_types[idx]
+            assert (
+                sig_type == in_type
+            ), "Export produced a graph with mismatched type signature {sig_type} vs expected {in_type} for arg {idx}"
+
+    if len(in_sig.parameters) == len(out_sig.parameters):
+        assert_equal_params(input_types)
+    else:
+        from functorch._src.aot_autograd import pytree
```
Comment (Chillee) [inline torchdynamo/eval_frame.py:293]: pytree is also in `torch.utils._pytree`

Comment (voznesenskym) [reply torchdynamo/eval_frame.py:293]: Even better, thanks :)

File: torchdynamo/eval_frame.py [hunk 2]
```diff
@@ -279,16 +279,28 @@ def dynamo_normalization_capturing_compiler(
...
+        flat_args, _ = pytree.tree_flatten(args, kwargs)
+        for arg in flat_args:
+            flat_input_types.append(arg.__class__)
+
+        assert len(flat_input_types) == len(
+            out_sig.parameters
+        ), "Flattened inputs length must match out signature parameter lengths."
+        assert_equal_params(flat_input_types)
 
     return (graph, out_guards)
```
Comment (jansel) [inline torchdynamo/eval_frame.py:305]: How would the user call the flattened graph i we don't return the pytree?

Comment (voznesenskym) [reply torchdynamo/eval_frame.py:305]: Good question. Where I integrate it now, we call pytree on the caller side, and pass in the flattened args again.

**Review by jansel — APPROVED** [review]

**Pull request merged**